--- /home/mouse/RPM/ali/shannon-direct-io-pcie-driver-3.4.1/shannon_device.c	2024-11-15 08:53:35.755376640 +0200
+++ shannon_device.c	2021-06-13 00:01:38.000000000 +0300
@@ -8,6 +8,9 @@
 #include <linux/jiffies.h>
 #include <linux/module.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 7, 0)
+#include <linux/part_stat.h>
+#endif
 #ifdef CONFIG_SUSE_PRODUCT_CODE
 #include <linux/suse_version.h>
 #endif
@@ -80,13 +83,9 @@ shannon_device_t *shannon_device_create(
 	va_end(vargs);
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 27)
-
 	return device_create((struct class *)class, (struct device *)parent, devt, drvdata, namebuf);
-
 #else
-
 	return device_create((struct class *)class, (struct device *)parent, devt, namebuf);
-
 #endif
 }
 
@@ -116,6 +115,201 @@ void shannon_unregister_chrdev_region(sh
 	unregister_chrdev_region(from, count);
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 8, 0)
+int shannon_disk_in_flight(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+	struct hd_struct *part = &gd->part0;
+	int cpu;
+	long sum = 0;
+
+	for_each_possible_cpu(cpu) {
+		sum += part_stat_local_read_cpu(part, in_flight[0], cpu);
+		sum += part_stat_local_read_cpu(part, in_flight[1], cpu);
+	}
+
+	return sum != 0;
+}
+
+unsigned long shannon_start_io_acct(shannon_gendisk_t *gdt,
+			   shannon_request_queue_t *queue,
+			   shannon_bio_t *p)
+{
+	struct bio *bio = (struct bio *) p;
+
+	return bio_start_io_acct(bio);
+}
+
+void shannon_end_io_acct(shannon_gendisk_t *gdt,
+			 shannon_request_queue_t *queue,
+			 shannon_bio_t *p,
+			 unsigned long start_time)
+{
+	struct bio *bio = (struct bio *) p;
+
+	bio_end_io_acct(bio, start_time);
+}
+
+unsigned long shannon_read_sectors(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+
+	if (gd == NULL)
+		return 0;
+
+	return part_stat_read(&gd->part0, sectors[READ]);
+}
+
+unsigned long shannon_write_sectors(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+
+	if (gd == NULL)
+		return 0;
+
+	return part_stat_read(&gd->part0, sectors[WRITE]);
+}
+
+unsigned long shannon_read_ios(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+
+	if (gd == NULL)
+		return 0;
+
+	return part_stat_read(&gd->part0, ios[READ]);
+}
+
+unsigned long shannon_write_ios(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+
+	if (gd == NULL)
+		return 0;
+
+	return part_stat_read(&gd->part0, ios[WRITE]);
+}
+
+unsigned long shannon_read_msecs(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+
+	if (gd == NULL)
+		return 0;
+
+	return part_stat_read(&gd->part0, nsecs[READ])/1000000;
+}
+
+unsigned long shannon_write_msecs(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+
+	if (gd == NULL)
+		return 0;
+
+	return part_stat_read(&gd->part0, nsecs[WRITE])/1000000;
+}
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(5, 7, 0)
+int shannon_disk_in_flight(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+	struct hd_struct *part = &gd->part0;
+	int cpu;
+	long sum = 0;
+
+	for_each_possible_cpu(cpu) {
+		sum += part_stat_local_read_cpu(part, in_flight[0], cpu);
+		sum += part_stat_local_read_cpu(part, in_flight[1], cpu);
+	}
+
+	return sum != 0;
+}
+
+unsigned long shannon_start_io_acct(shannon_gendisk_t *gdt,
+			   shannon_request_queue_t *queue,
+			   shannon_bio_t *p)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+	struct request_queue *q = (struct request_queue *) queue;
+	struct bio *bio = (struct bio *) p;
+	unsigned long now = get_jiffies();
+
+	generic_start_io_acct(q, bio_op(bio), bio_sectors(bio), &gd->part0);
+	return now;
+}
+
+void shannon_end_io_acct(shannon_gendisk_t *gdt,
+			 shannon_request_queue_t *queue,
+			 shannon_bio_t *p,
+			 unsigned long start_time)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+	struct request_queue *q = (struct request_queue *) queue;
+	struct bio *bio = (struct bio *) p;
+
+	generic_end_io_acct(q, bio_op(bio), &gd->part0, start_time);
+}
+
+unsigned long shannon_read_sectors(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+
+	if (gd == NULL)
+		return 0;
+
+	return part_stat_read(&gd->part0, sectors[READ]);
+}
+
+unsigned long shannon_write_sectors(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+
+	if (gd == NULL)
+		return 0;
+
+	return part_stat_read(&gd->part0, sectors[WRITE]);
+}
+
+unsigned long shannon_read_ios(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+
+	if (gd == NULL)
+		return 0;
+
+	return part_stat_read(&gd->part0, ios[READ]);
+}
+
+unsigned long shannon_write_ios(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+
+	if (gd == NULL)
+		return 0;
+
+	return part_stat_read(&gd->part0, ios[WRITE]);
+}
+
+unsigned long shannon_read_msecs(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+
+	if (gd == NULL)
+		return 0;
+
+	return part_stat_read(&gd->part0, nsecs[READ])/1000000;
+}
+
+unsigned long shannon_write_msecs(shannon_gendisk_t *gdt)
+{
+	struct gendisk *gd = (struct gendisk *)gdt;
+
+	if (gd == NULL)
+		return 0;
+
+	return part_stat_read(&gd->part0, nsecs[WRITE])/1000000;
+}
+#else
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 28)
 
@@ -151,16 +345,15 @@ int shannon_disk_in_flight(shannon_gendi
 
 static inline void shannon_part_inc_in_flight(struct hd_struct *part, int rw)
 {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) ||	\
-	(defined(SHANNON_RHEL_RELEASE_OVER_8_0))
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) || defined(SHANNON_RHEL_RELEASE_OVER_8_0)
 	part_stat_local_inc(part, in_flight[rw]);
 #else
 	atomic_inc((atomic_t *)(&part->in_flight[rw]));
 #endif
+
 	if (part->partno)
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) ||	\
-	(defined(SHANNON_RHEL_RELEASE_OVER_8_0))
-	part_stat_local_inc(&part_to_disk(part)->part0, in_flight[rw]);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) || defined(SHANNON_RHEL_RELEASE_OVER_8_0)
+		part_stat_local_inc(&part_to_disk(part)->part0, in_flight[rw]);
 #else
 		atomic_inc((atomic_t *)(&part_to_disk(part)->part0.in_flight[rw]));
 #endif
@@ -168,16 +361,15 @@ static inline void shannon_part_inc_in_f
 
 static inline void shannon_part_dec_in_flight(struct hd_struct *part, int rw)
 {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) ||	\
-	(defined(SHANNON_RHEL_RELEASE_OVER_8_0))
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) || defined(SHANNON_RHEL_RELEASE_OVER_8_0)
 	part_stat_local_dec(part, in_flight[rw]);
 #else
 	atomic_dec((atomic_t *)(&part->in_flight[rw]));
 #endif
+
 	if (part->partno)
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) ||	\
-	(defined(SHANNON_RHEL_RELEASE_OVER_8_0))
-	part_stat_local_dec(&part_to_disk(part)->part0, in_flight[rw]);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) || defined(SHANNON_RHEL_RELEASE_OVER_8_0)
+		part_stat_local_dec(&part_to_disk(part)->part0, in_flight[rw]);
 #else
 		atomic_dec((atomic_t *)(&part_to_disk(part)->part0.in_flight[rw]));
 #endif
@@ -187,34 +379,40 @@ int shannon_disk_in_flight(shannon_gendi
 {
 	struct gendisk *gd = (struct gendisk *)gdt;
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) ||	\
-	(defined(SHANNON_RHEL_RELEASE_OVER_8_0))
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) || defined(SHANNON_RHEL_RELEASE_OVER_8_0)
 	int cpu;
 	unsigned int inflight;
 #endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 14, 0)
 
 #ifdef SHANNON_ON_RHEL
+
 #ifdef SHANNON_RHEL_RELEASE_OVER_7_6
 	return atomic_read(&gd->part0.in_flight[0]) + atomic_read(&gd->part0.in_flight[1]);
 #else
 	return part_in_flight(&gd->part0);
 #endif
+
 #else
+
 #ifdef SHANNON_ON_SUSE
+
 #ifdef SHANNON_SUSE_RELEASE_BELOW_1_12_5
 	return part_in_flight(&gd->part0);
 #else
 	return atomic_read(&gd->part0.in_flight[0]) + atomic_read(&gd->part0.in_flight[1]);
 #endif
+
 #else
 	return part_in_flight(&gd->part0);
 #endif
+
 #endif
 
 #else
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) ||	\
-	(defined(SHANNON_RHEL_RELEASE_OVER_8_0))
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) || defined(SHANNON_RHEL_RELEASE_OVER_8_0)
 	inflight = 0;
 	for_each_possible_cpu(cpu) {
 		inflight += part_stat_local_read_cpu(&gd->part0, in_flight[0], cpu) + \
@@ -227,13 +425,13 @@ int shannon_disk_in_flight(shannon_gendi
 #else
 	return atomic_read(&gd->part0.in_flight[0]) + atomic_read(&gd->part0.in_flight[1]);
 #endif
+
 #endif
 }
 
 #endif
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) ||	\
-	(defined(SHANNON_RHEL_RELEASE_OVER_8_0))
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) || defined(SHANNON_RHEL_RELEASE_OVER_8_0)
 void shannon_update_io_ticks(struct hd_struct *part, unsigned long now)
 {
         unsigned long stamp;
@@ -251,53 +449,78 @@ again:
 }
 #endif
 
-void shannon_start_io_acct(shannon_gendisk_t *gdt, shannon_bio_t *p)
+unsigned long shannon_start_io_acct(shannon_gendisk_t *gdt,
+			   shannon_request_queue_t *queue,
+			   shannon_bio_t *p)
 {
 	struct bio *bio = (struct bio *)p;
+	unsigned long now = get_jiffies();
+
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 28)
-	int cpu, rw;
+	int rw;
+#if !defined(SHANNON_RHEL_RELEASE_OVER_8_4)
+	int cpu;
+#endif
 	struct gendisk *gd = (struct gendisk *)gdt;
 	struct hd_struct *part;
 
 	rw = shannon_bio_data_dir(bio);
+#if defined(SHANNON_RHEL_RELEASE_OVER_8_4)
+	part_stat_lock();
+#else
 	cpu = part_stat_lock();
+#endif
+
+#if defined(SHANNON_RHEL_RELEASE_OVER_8_4)
+	part = &gd->part0;
+#else
 	part = disk_map_sector_rcu(gd, get_bi_sector(bio));
+#endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 14, 0)
 
 #ifdef SHANNON_ON_RHEL
+
 #ifdef SHANNON_RHEL_RELEASE_OVER_7_6
 	part_round_stats(gd->queue, cpu, part);
 #else
 	part_round_stats(cpu, part);
 #endif
+
 #else
+
 #ifdef SHANNON_ON_SUSE
+
 #ifdef SHANNON_SUSE_RELEASE_BELOW_1_12_5
 	part_round_stats(cpu, part);
 #else
 	part_round_stats(gd->queue, cpu, part);
 #endif
+
 #else
 	part_round_stats(cpu, part);
 #endif
+
 #endif
 
 #else
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) ||	\
-	(defined(SHANNON_RHEL_RELEASE_OVER_8_0))
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) || defined(SHANNON_RHEL_RELEASE_OVER_8_0)
 	shannon_update_io_ticks(part, get_jiffies());
 #else
 	part_round_stats(gd->queue, cpu, part);
 #endif
+
 #endif
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) ||	\
-	(defined(SHANNON_RHEL_RELEASE_OVER_8_0))
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) || defined(SHANNON_RHEL_RELEASE_OVER_8_0)
 	part_stat_inc(part, ios[rw]);
 	part_stat_add(part, sectors[rw], bio_sectors(bio));
 #else
 	part_stat_inc(cpu, part, ios[rw]);
 	part_stat_add(cpu, part, sectors[rw], bio_sectors(bio));
 #endif
+
 	shannon_part_inc_in_flight(part, rw);
 	part_stat_unlock();
 #else
@@ -312,58 +535,85 @@ void shannon_start_io_acct(shannon_gendi
 	atomic_inc((atomic_t *)(&gd->in_flight));
 	preempt_enable();
 #endif
-
+	return now;
 }
 
-void shannon_end_io_acct(shannon_gendisk_t *gdt, shannon_bio_t *p, unsigned long duration)
+void shannon_end_io_acct(shannon_gendisk_t *gdt,
+			 shannon_request_queue_t *queue,
+			 shannon_bio_t *p,
+			 unsigned long start_time)
 {
 	struct bio *bio = (struct bio *)p;
+	unsigned int duration = get_jiffies() - start_time;
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 28)
-	int cpu, rw;
+	int rw;
+#if !defined(SHANNON_RHEL_RELEASE_OVER_8_4)
+	int cpu;
+#endif
 	struct gendisk *gd = (struct gendisk *)gdt;
 	struct hd_struct *part;
 
 	rw = shannon_bio_data_dir(bio);
+#if defined(SHANNON_RHEL_RELEASE_OVER_8_4)
+	part_stat_lock();
+#else
 	cpu = part_stat_lock();
+#endif
+
+#if defined(SHANNON_RHEL_RELEASE_OVER_8_4)
+	part = &gd->part0;
+#else
 	part = disk_map_sector_rcu(gd, get_bi_sector(bio));
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 19, 0) ||	\
-	(defined(SHANNON_RHEL_RELEASE_OVER_8_0))
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) ||	\
-	(defined(SHANNON_RHEL_RELEASE_OVER_8_0))
+#endif
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 19, 0) || defined(SHANNON_RHEL_RELEASE_OVER_8_0)
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0) || defined(SHANNON_RHEL_RELEASE_OVER_8_0)
 	shannon_update_io_ticks(part, get_jiffies());
 	part_stat_add(part, nsecs[rw], shannon_jiffies_to_usecs(duration) * 1000);
+#if !defined(SHANNON_RHEL_RELEASE_OVER_8_4)
 	part_stat_add(part, time_in_queue, duration);
+#endif
 #else
 	part_stat_add(cpu, part, nsecs[rw], shannon_jiffies_to_usecs(duration) * 1000);
 #endif
+
 #else
 	part_stat_add(cpu, part, ticks[rw], duration);
 #endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 14, 0)
 
 #ifdef SHANNON_ON_RHEL
+
 #ifdef SHANNON_RHEL_RELEASE_OVER_7_6
 	part_round_stats(gd->queue, cpu, part);
 #else
 	part_round_stats(cpu, part);
 #endif
+
 #else
+
 #ifdef SHANNON_ON_SUSE
+
 #ifdef SHANNON_SUSE_RELEASE_BELOW_1_12_5
 	part_round_stats(cpu, part);
 #else
 	part_round_stats(gd->queue, cpu, part);
 #endif
+
 #else
 	part_round_stats(cpu, part);
 #endif
+
 #endif
 
 #else
-#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 0, 0) &&	\
-	(defined(SHANNON_RHEL_RELEASE_BELOW_8_0))
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 0, 0) &&	!defined(SHANNON_RHEL_RELEASE_OVER_8_0)
 	part_round_stats(gd->queue, cpu, part);
 #endif
+
 #endif
 	shannon_part_dec_in_flight(part, rw);
 	part_stat_unlock();
@@ -444,12 +694,13 @@ unsigned long shannon_read_msecs(shannon
 		return 0;
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 28)
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 19, 0) ||	\
-	(defined(SHANNON_RHEL_RELEASE_OVER_8_0))
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 19, 0) || defined(SHANNON_RHEL_RELEASE_OVER_8_0)
 	return part_stat_read(&gd->part0, nsecs[READ])/1000000;
 #else
 	return jiffies_to_msecs(part_stat_read(&gd->part0, ticks[READ]));
 #endif
+
 #else
 	return jiffies_to_msecs(disk_stat_read(gd, ticks[READ]));
 #endif
@@ -463,13 +714,15 @@ unsigned long shannon_write_msecs(shanno
 		return 0;
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 28)
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 19, 0) ||	\
-	(defined(SHANNON_RHEL_RELEASE_OVER_8_0))
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 19, 0) || defined(SHANNON_RHEL_RELEASE_OVER_8_0)
 	return part_stat_read(&gd->part0, nsecs[WRITE])/1000000;
 #else
 	return jiffies_to_msecs(part_stat_read(&gd->part0, ticks[WRITE]));
 #endif
+
 #else
 	return jiffies_to_msecs(disk_stat_read(gd, ticks[WRITE]));
 #endif
 }
+#endif
